src/hipscat/pixel_tree/pixel_alignment.py

```diff
@@ -0,0 +1,176 @@
+
+    PRIMARY_ORDER_COLUMN_NAME = "primary_Norder"
+    PRIMARY_PIXEL_COLUMN_NAME = "primary_Npix"
+    JOIN_ORDER_COLUMN_NAME = "join_Norder"
+    JOIN_PIXEL_COLUMN_NAME = "join_Npix"
+    ALIGNED_ORDER_COLUMN_NAME = "aligned_Norder"
+    ALIGNED_PIXEL_COLUMN_NAME = "aligned_Npix"
+
+    def __init__(
+            self,
+            aligned_tree: PixelTree,
+            pixel_mapping: pd.DataFrame,
+            alignment_type: PixelAlignmentType
+    ) -> None:
+        self.pixel_tree = aligned_tree
+        self.pixel_mapping = pixel_mapping
+        self.alignment_type = alignment_type
+
+
+# pylint: disable=R0912
+def align_trees(
+        left: PixelTree,
+        right: PixelTree,
+        alignment_type: PixelAlignmentType = PixelAlignmentType.INNER
+) -> PixelAlignment:
+    """Generate a `PixelAlignment` object from two pixel trees
+
+    A `PixelAlignment` represents how two pixel trees align with eachother, meaning which pixels
+    match or overlap between the catalogs, and includes a new tree with the smallest pixels from
+    each tree
+
+    Args:
+        left: The left tree to align
+        right: The right tree to align
+        alignment_type: The type of alignment describing how to handle nodes which exist in one tree
+            but not the other. Options are:
+            inner - only use pixels that appear in both catalogs
+            left - use all pixels that appear in the left catalog and any overlapping from the right
+            right - use all pixels that appear in the right catalog and any overlapping from the
+                left
+            outer - use all pixels from both catalogs
+    Returns:
+        The `PixelAlignment` object with the alignment from the two trees
+    """
+
+    tree_builder = PixelTreeBuilder()
+    pixels_to_search = _get_children_pixels_from_trees([left, right],
+                                                            left.root_pixel.pixel)
+
+    while len(pixels_to_search) > 0:
+        search_pixel = pixels_to_search.pop(0)
+        if search_pixel in left and search_pixel in right:
+            left_node = left[search_pixel]
+            right_node = right[search_pixel]
+            if left_node.node_type == right_node.node_type:
+                if left_node.node_type == PixelNodeType.LEAF:
+                    tree_builder.create_node_and_parent_if_not_exist(search_pixel,
+                                                                     PixelNodeType.LEAF)
+                else:
+                    pixels_to_search += _get_children_pixels_from_trees([left, right], search_pixel)
+            else:
+                if left_node.node_type == PixelNodeType.LEAF:
+                    tree_with_leaf_node = "left"
+                    inner_node = right_node
+                else:
+                    tree_with_leaf_node = "right"
+                    inner_node = left_node
+                if _should_include_all_pixels_from_tree(tree_with_leaf_node, alignment_type):
```

Comment:
I'm having trouble following this IF block. If we want to include all pixels from the tree, I would expect to label the search node inner and add all descendants from the other tree. Otherwise, I would expect to add the leaf to the tree and not split it.

Can you clarify where my logic is off? Maybe it would be best to go over this at a whiteboard.

Reply:
This makes sense given the design document. The first case is adding the node as a leaf and splitting to include all descendants. The second case only includes the actual descendants of the inner node. Correct?

Reply:
Yep, that's right!